watch/watch.go

```diff
@@ -0,0 +1,360 @@
+	ack  signal // closed by the sending goroutine to tell the watcher it has stopped sending query results.
+}
+
+// New creates and returns a new Watcher on the storage.Store s using the ast.Compiler c.
+// Once a compiler is provided to create a Watcher, it must not be modified, or else
+// the results produced by the Watcher are undefined.
+//
+// The Watcher will already be started by New.
+//
+// The ID provided must be unique to the storage.Store.
+func New(ctx context.Context, s storage.Store, c *ast.Compiler, id string, txn storage.Transaction) (*Watcher, error) {
+	w := create(ctx, s, c, id)
+	return w, w.Start(txn)
+}
+
+// Start starts the watcher listening for changes to queries registered on it.
+func (w *Watcher) Start(txn storage.Transaction) error {
+	return w.store.Register(w.ctx, txn, w.id, storage.TriggerConfig{OnCommit: w.notify})
+}
+
+// Query registers a watch on the provided Rego query. Whenever changes are made to a
+// base or virtual document that the query depends on, an Event describing the new result
+// of the query will be sent along the returned channel. The provided id string may be
+// used to end the watch on the query.
+//
+// WatchQuery will return an error if registering the watch fails for any reason.
+func (w *Watcher) Query(query, id string) (<-chan Event, error) {
+	out := make(chan Event)
+	if err := w.query(query, id, out); err != nil {
+		close(out)
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// EndQuery ends the watch on the query associated with id. It will close the channel
+// that was delivering notifications for the given ID. This may happen before or after
+// EndQuery returns.
+func (w *Watcher) EndQuery(id string) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	return w.endQuery(id, false)
+}
+
+// Close ends the watches on all queries this Watcher has. If closing the Watcher fails,
+// no changes will be made to it.
+//
+// Further attempts to register or end watches will result in an error after Close() is
+// called.
+func (w *Watcher) Close(txn storage.Transaction) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	if w.closed {
+		return errors.New("watcher already closed")
+	}
+
+	w.close(txn, false)
+	return nil
+}
+
+// Migrate creates a new Watcher with the same watches as w, but using the new
+// ast.Compiler. Like when creating a Watcher with New, the provided compiler must not
+// be modified after being passed to Migrate, or else behavior is undefined.
+//
+// Migrate will not start the new Watcher, it must be started with the Start function
+// before it will deliver notifications.
+//
+// After Migrate returns, the old watcher will be closed, and the new will be ready for
+// use. All channels from the old watcher will still be active, via the returned Watcher.
+//
+// If an error occurs creating the new Watcher, the state of the old Watcher will not be
+// changed.
+func (w *Watcher) Migrate(c *ast.Compiler, txn storage.Transaction) (*Watcher, error) {
```

Comment:
It would be good to include an example test for this. I'm wondering if this interface can be simplified at all. I'd like to reduce the amount of lifecycle management that the caller has to do.